File: 2021/day01/Main.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE TypeApplications #-}

input :: String -> IO [Int]
input name = do
    (fmap (read @Int) . lines) <$> readFile name

calcIncreases1 :: [Int] -> Int
calcIncreases1 = \case
    [] -> 0
    head : tail ->
        snd $ foldl calc (head, 0) tail
  where
    calc (prev, count) val =
        ( val
        , if prev < val
            then count + 1
            else count
        )

calcIncreases2 :: [Int] -> Int
calcIncreases2 = \case
    h1 : h2 : h3 : tail ->
        snd $ foldl calc ((h1, h2, h3), 0) tail
    _ -> 0
  where
    calc ((prev@(p1, p2, p3)), count) val =
        let cur = (p2, p3, val)
         in ( cur
            , if sum prev < sum cur
                then count + 1
                else count
            )
    sum (i1, i2, i3) = i1 + i2 + i3

main :: IO ()
main = do
    vals <- input "day01/input.txt"
    print $ calcIncreases1 vals
    print $ calcIncreases2 vals

File: 2021/day01/main.rs
use std::fmt::Debug;
use std::fs::File;
use std::io::{BufRead, BufReader, Lines};

fn calc_increases1(values: &Vec<u32>) -> u32 {
    let mut prev = values[0];
    let mut count: u32 = 0;

    for val in values {
        if prev < *val {
            count += 1;
        }
        prev = *val;
    }
    count
}

fn sum(v1: u32, v2: u32, v3: u32) -> u32 {
    v1 + v2 + v3
}

fn calc_increases2(values: &Vec<u32>) -> u32 {
    let mut p1 = values[0];
    let mut p2 = values[1];
    let mut p3 = values[2];
    let mut count: u32 = 0;
    let mut index = -1;

    for val in values {
        index += 1;
        if index < 3 {
            continue;
        }
        if sum(p1, p2, p3) < sum(p2, p3, *val) {
            count += 1;
        }

        p1 = p2;
        p2 = p3;
        p3 = *val;
    }
    count
}

fn parse_u32<T: Debug>(res: Result<String, T>) -> u32 {
    res.unwrap().parse().unwrap()
}

fn main() {
    let file = File::open("day01/input.txt").unwrap();
    let lines: Lines<BufReader<File>> = BufReader::new(file).lines();

    let mut vals: Vec<u32> = Vec::new();

    for (_index, line) in lines.enumerate() {
        vals.push(parse_u32(line));
    }

    println!("{}", calc_increases1(&vals));
    println!("{}", calc_increases2(&vals));
}
